### PR TITLE
CiviCRM API, lookup state_province_id options based on country parame…

### PR DESCRIPTION
**Pull request by jackrabbithanna**
…ter if present, or default country

Overview
----------------------------------------
See issue: https://lab.civicrm.org/dev/core/issues/725

Before
----------------------------------------
CiviCRM Address API incorrectly sets state_province_id for states whose name/abbreviation exist in multiple countries.

After
----------------------------------------
CiviCRM Address API correctly sets state_province_id for states whose name/abbreviation exist in multiple countries.

Technical Details
----------------------------------------
The API does not pass parameters to the functions/methods responsible for building the list of valid options for the state_province_id field, instead it produces the entire list of states in the system. API validation does not catch this situation. Then if a state name or abbreviation is used by multiple countries, the first option found is used, instead of the right one. This condition could also exist for counties. Examples are "Maryland" and "Montana" . See issue for code example.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@jackrabbithanna I'm OK with the slightly field-specific hackery of this if we can get a unit test on it - @colemanw can you see any issues with the change to Address::buildOptions? I'm not too sure why that context check is there

**Comment by jackrabbithanna** [comment]
@eileenmcnaughton do you think this test failure is related? It's hard for me to see how it could be.

**Comment by jackrabbithanna** [comment]
Probably I should put the new test in tests/phpunit/api/v3/AddressTest.php ...I'll write one.

**Comment by jackrabbithanna** [comment]
Created testCreateAddressStateProvinceIDCorrectForCountry() in tests/phpunit/api/v3/AddressTest.php

File: CRM/Core/BAO/Address.php [hunk 1]
```diff
@@ -1320,7 +1320,7 @@ public static function buildOptions($fieldName, $context = NULL, $props = array(
             $props['country_id'] = $config->defaultContactCountry;
           }
         }
-        if (!empty($props['country_id']) && $context !== 'validate') {
+        if (!empty($props['country_id'])) {
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/Address.php]: So this is the line I think is risky - the single field handling below is a bit hacky but mitigated by the unit test IMHO so I'm OK with it 

I found where this was added - @colemanw was fixing a webform bug

https://issues.civicrm.org/jira/browse/CRM-15215

**Comment by eileenmcnaughton** [comment]
@colemanw I'd really like your thoughts on this - my feeling is that if we could order the return results to put the preferred country first it would be safer but that is a bit tricky

**Comment by eileenmcnaughton** [comment]
@colemanw assigning this to you as I want your input since it undoes a 'fix' you did earlier

**Comment by colemanw** [comment]
Ok after reviewing this I've pushed up 2 changes

1. Add back the condition that fixed the webform bug, but more specific this time. We don't apply the default country when validating states but we will respect a specified country.
2. Make the test cover both Marylands so that we don't get an accidental pass for one or the other.

**Comment by eileenmcnaughton** [comment]
Yay it passed - merging - what's your next api edge case @jackrabbithanna ?

**Comment by eileenmcnaughton** [comment]
Just noting this  caused a regression picked up through other automated process - I think the CT is working on resolving at this stage
